earth_enterprise/src/scons/khEnvironment.py

```diff
@@ -159,6 +162,192 @@ def EmitBuildDateStrfunc(target, build_date):
+
+    return final
+  
+
+def CheckGitAvailable():
+    """Try the most basic of git commands, to see if there is
+       currently any access to a repository."""
+    
+    try:
+        subprocess.check_output(['git', 'status'])
+    except Exception:
+        return False
+    
+    return True
+
+
+def CheckDirtyRepository():
+    """Check to see if the repository is not in a cleanly committed state."""
+    
+    str = subprocess.check_output(['git', 'status', '--porcelain'])
+    
+    # Ignore version.txt for this purpose, as a build may modify the file
+    # and lead to an erroneous interpretation on repeated consecutive builds.
+    if (str == " M earth_enterprise/src/version.txt\n"):
+        return False
+    
+    return (len(str) > 0)
+    
+
+def ReadBackupVersionFile(target):
+    """There should be a file checked in with the latest version
+       information available; if git isn't available to provide
+       information, then use this file instead."""
+    
+    try:
+        fp = open(target, 'r')
+        line = fp.readline()
+        fp.close()
+        
+        return line
+    
+    except Exception:
+        return "Version Error"
+
+
+def GitGeneratedLongVersion():
+    """Take the raw information parsed by git, and use it to
+       generate an appropriate version string for GEE."""
+    try:
+        raw = subprocess.check_output(['git', 'describe', '--tags',
+                                        '--match', '[0-9]*\.[0-9]*\.[0-9]*\-*'])
+        raw = raw.rstrip()
+    except Exception:
+        return "Version Error"
+
+    # Grab the datestamp.
+    date = datetime.utcnow().strftime("%Y%m%d%H%M")
+
+    # If this condition hits, then we are currently on a tagged commit.
+    if (len(raw.split("-")) < 4):
+        if CheckDirtyRepository():
+            return '.'.join([raw, date])
+        return raw
+
+    # Tear apart the information in the version string.
+    components = ParseRawVersionString(raw)
+  
+    # Determine how to update, since we are *not* on tagged commit.
+    if components['isFinal']:
+        components['patch'] = 0
+        components['patchType'] = "alpha"
+        components['revision'] = components['revision'] + 1
+    else:
+        components['patch'] = components['patch'] + 1
+    
+    # Rebuild.
+    base = '.'.join([str(components[x]) for x in ("major", "minor", "revision")])
+    patch = '.'.join([str(components["patch"]), components["patchType"], date])
+    if not CheckDirtyRepository():
+        patch = '.'.join([patch, components['hash']])
+    
+    return '-'.join([base, patch])
+
+
+def ParseRawVersionString(raw):
+    """Break apart a raw version string into its various components,
+    and return those entries via a dictionary."""
+
+    components = { }    
+    rawComponents = raw.split("-")
+    
+    base = rawComponents[0]
+    patchRaw = rawComponents[1]
+    components['numCommits'] = rawComponents[2]
+    components['hash'] = rawComponents[3]
```

Comment:
Q: Are we including a hash as part of a Git tag?  Doesn't Git already know what commit it's at?

Reply:
No, none of the git tags should have a hash.